### ISSUE TITLE
Allow for tag-specific schemes

### ISSUE THREAD

**Issue by cwill747** [punkave/sanitize-html#50]
Would like the ability to specify what schemes are supported on a per-tag basis. I could submit a pull request if you are interested.

For example:
```javascript
sanitizeHtml.defaults = {
  allowedTags: ['h3', 'h4', 'h5', 'h6', 'blockquote', 'p', 'a', 'ul', 'ol', 'nl', 'li', 'b', 'i', 'strong', 'em', 'strike', 'code', 'hr', 'br', 'div', 'table', 'thead', 'caption', 'tbody', 'tr', 'th', 'td', 'pre'],
  allowedAttributes: {
    a: ['href', 'name', 'target'],
    // We don't currently allow img itself by default, but this
    // would make sense if we did
    img: ['src']
  },
  // Lots of these won't come up by default because we don't allow them
  selfClosing: ['img', 'br', 'hr', 'area', 'base', 'basefont', 'input', 'link', 'meta'],
  // URL schemes we permit
  allowedSchemes: {
    a: ['http'],
    sanitizeDefault: ['http', 'https', 'ftp', 'mailto']
  }
};
```

**Comment by boutell** [comment]
Got any examples of things that would be safe for one tag but not for
another?

On Mon, Apr 27, 2015 at 2:57 PM, Cameron Will <notifications@github.com>
wrote:

> Would like the ability to specify what schemes are supported on a per-tag
> basis. I could submit a pull request if you are interested.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/punkave/sanitize-html/issues/50>.
>



-- 


*THOMAS BOUTELL, *DEV & OPS
P'UNK AVENUE | (215) 755-1330  |  punkave.com

**Comment by cwill747** [comment]
This is more just of a personal preference. For example, we use this to sanitize wiki text put in by a user. The only place we want to allow any sort of scheme is in the <a> tag, and we want the href to allow https:// only. We don't, however, want any http:// or https:// tags allowed in the src= field of the <img> tag, but only the data:// scheme.

**Comment by boutell** [comment]
Interesting. I don't see any reason why this couldn't be accommodated as a
backwards-compatible pull request.

On Tue, Apr 28, 2015 at 11:09 AM, Cameron Will <notifications@github.com>
wrote:

> This is more just of a personal preference. For example, we use this to
> sanitize wiki text put in by a user. The only place we want to allow any
> sort of scheme is in the tag, and we want the href to allow https://
> only. We don't, however, want any http:// or https:// tags allowed in the
> src= field of the tag, but only the data:// scheme.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/punkave/sanitize-html/issues/50#issuecomment-97100102>
> .
>



-- 


*THOMAS BOUTELL, *DEV & OPS
P'UNK AVENUE | (215) 755-1330  |  punkave.com

**Comment by cwill747** [comment]
Good point on backwards-compatibility. I will make it so and submit a PR. Thanks!

**Comment by boutell** [comment]
I accepted and published this. I renamed it to allowedSchemesByTag, just to better match the naming convention I'm most likely to use in similar situations.

I rewrote your change in naughtyHref, because your original code would actually reject the scheme only if it was missing from both the tag-specific list and the global list. For this option to be effective, it needs to be set up so that if the tag-specific list exists, only that list is consulted.

I also changed your test, which was not sensitive enough to catch that bug. The "allowedSchemes" list only had http and the allowedSchemesByTag list had nothing for img, so it wasn't possible to tell from the test why the "data" attribute went away. There are now two asserts, one verifying that "data" is preserved when it is specifically allowed for img only, and one verifying that the src attribute is tossed out entirely when the list of schemes for img is empty.
